### ISSUE TITLE
borg info incorrect output on ppc64

### ISSUE THREAD

**Issue by verygreen** [borgbackup/borg#886]
Using ppc64 Fedora22 with self-compiled python-msgpack 0.4.7 backing up to x86 box and self-compiled borg 1.0.0 (with pip), I have this:

```
                       Original size      Compressed size    Deduplicated size
This archive:               15.07 GB              8.63 GB             62.27 MB
All archives:                4.50 EB             -7.23 EB            -12.31 TB

                       Unique chunks         Total chunks
Chunk index:                  274262       10289249058816
```

Meanwhile same archive accessed from the x86 box locally yields:
```
                       Original size      Compressed size    Deduplicated size
This archive:               15.07 GB              8.63 GB             62.27 MB
All archives:               29.93 GB             17.22 GB              8.18 GB

                       Unique chunks         Total chunks
Chunk index:                  274262               618151
```

Apparently the problem does not have anything to do with cross-endianness communication too as it's reproduceable locally as well:

```
[root@powermac ~]# borg init --encryption none /tmp/backup
[root@powermac ~]# borg create /tmp/backup::first /etc/
[root@powermac ~]# touch /etc/file
[root@powermac ~]# borg create /tmp/backup::second /etc/
[root@powermac ~]# borg info /tmp/backup::second
Name: second
Fingerprint: fe9ba71131e7d07915369259c01a7a448f09fd727c9ed91121e399c25c19eaa6
Hostname: powermac.localnet
Username: root
Time (start): Mon, 2016-04-11 15:04:50
Time (end):   Mon, 2016-04-11 15:04:51
Command line: /usr/bin/borg create /tmp/backup::second /etc/
Number of files: 2013

                       Original size      Compressed size    Deduplicated size
This archive:               25.58 MB             25.59 MB            498.40 kB
All archives:                6.86 EB              8.11 EB            331.42 GB

                       Unique chunks         Total chunks
Chunk index:                    1965          67813507072
```

**Comment by verygreen** [comment]
This looks like some endianness error somewhere (forgot to swab some stuff)?
```
(gdb) p/x 618151
$3 = 0x96ea7
(gdb) p/x 10289249058816
$4 = 0x95ba7090000
(gdb) 
(gdb) p/x 67813507072
$1 = 0xfca000000
(gdb) p 0xfca
$2 = 4042
```

**Comment by verygreen** [comment]
Ah, I think I have an even better demonstration:

```
[root@powermac ~]# borg init --encryption none /tmp/backup
[root@powermac ~]# touch /tmp/file
[root@powermac ~]# borg create /tmp/backup::first /tmp/file
[root@powermac ~]# borg info /tmp/backup::first
Name: first
Fingerprint: 4480529f9d5ec35749063f5894752edc44996c31c2dc126e0cf953e555d3d42f
Hostname: powermac.localnet
Username: root
Time (start): Mon, 2016-04-11 15:18:58
Time (end):   Mon, 2016-04-11 15:18:58
Command line: /usr/bin/borg create /tmp/backup::first /tmp/file
Number of files: 1

                       Original size      Compressed size    Deduplicated size
This archive:                  402 B                408 B                408 B
All archives:              -30.96 PB            -29.27 PB             -1.74 GB

                       Unique chunks         Total chunks
Chunk index:                       2             33554432

(gdb) p/x 33554432
$5 = 0x2000000
```

**Comment by enkore** [comment]
My stab at this: The data in the tree is stored in LE, but hashindex_summarize fails to convert LE to host-order. Can you try patching _hashindex.c like this:

from:
```c
static void
hashindex_summarize(HashIndex *index, long long *total_size, long long *total_csize,
                    long long *total_unique_size, long long *total_unique_csize,
                    long long *total_unique_chunks, long long *total_chunks)
{
    int64_t size = 0, csize = 0, unique_size = 0, unique_csize = 0, chunks = 0, unique_chunks = 0;
    const int32_t *values;
    void *key = NULL;

    while((key = hashindex_next_key(index, key))) {
        values = key + index->key_size;
        unique_chunks++;
        chunks += values[0];
        unique_size += values[1];
        unique_csize += values[2];
        size += (int64_t) values[0] * values[1];
        csize += (int64_t) values[0] * values[2];
    }
    *total_size = size;
    *total_csize = csize;
    *total_unique_size = unique_size;
    *total_unique_csize = unique_csize;
    *total_unique_chunks = unique_chunks;
    *total_chunks = chunks;
}
```

to

```c
    while((key = hashindex_next_key(index, key))) {
        values = key + index->key_size;
        unique_chunks++;
        chunks += _le32toh(values[0]);
        unique_size += _le32toh(values[1]);
        unique_csize += _le32toh(values[2]);
        size += (int64_t) _le32toh(values[0] * _le32toh(values[1]);
        csize += (int64_t) _le32toh(values[0]) * _le32toh(values[2]);
    }
```

I don't think anything here uses BE, so I can't test it myself.

**Comment by ThomasWaldmann** [comment]
maybe the code in _hashindex.c should be reviewed if there are other such issues.

**Comment by enkore** [comment]
_summarize is the only method that actually cares about the contents (I guess this was moved from Python or Cython there for performance reasons), otherwise there are few multi-byte values handled by it. num_entries and num_buckets look like the only ones and they are handled correctly.

**Comment by verygreen** [comment]
Ok, so with this patch:
```
--- a/borg/_hashindex.c
+++ b/borg/_hashindex.c
@@ -452,11 +452,11 @@ hashindex_summarize(HashIndex *index, long long *total_size, long long *total_cs
     while((key = hashindex_next_key(index, key))) {
         values = key + index->key_size;
         unique_chunks++;
-        chunks += values[0];
-        unique_size += values[1];
-        unique_csize += values[2];
-        size += (int64_t) values[0] * values[1];
-        csize += (int64_t) values[0] * values[2];
+        chunks += _le32toh(values[0]);
+        unique_size += _le32toh(values[1]);
+        unique_csize += _le32toh(values[2]);
+        size += (int64_t) _le32toh(values[0]) * _le32toh(values[1]);
+        csize += (int64_t) _le32toh(values[0]) * _le32toh(values[2]);
     }
     *total_size = size;
     *total_csize = csize;

```

The output seems to be correct now:
```
                       Original size      Compressed size    Deduplicated size
This archive:               15.06 GB              8.62 GB             26.81 MB
All archives:               44.99 GB             25.85 GB              8.21 GB

                       Unique chunks         Total chunks
Chunk index:                  278309               927140
```

**Comment by enkore** [comment]
Thanks for testing this :)

**Comment by verygreen** [comment]
so, I guess I must ask this now, how stable is the borg protocol wrt version changes. As it happens now, I have 1.0.0 on my server and 1.0.2+ on ppc client, is that ok and supported (is it supported in general to run different versions like this)?

Thanks.

**Comment by enkore** [comment]
That's fine. Generally speaking, incompatibilities to prior versions are noted in the changelog. The 1.x series will continue to use the same repository format, but it might happen inside 1.x that the RPC is changed in incompatible ways -- this will be clearly announced in the changelog if it happens.

**Comment by verygreen** [comment]
Ok, thanks.
I guess this ticket could be closed now.

**Comment by enkore** [comment]
Yep, the fix is merged into 1.0-maint and will be included in 1.0.2. It will also be merged back into master and be included in 1.1 etc.

**Comment by ThomasWaldmann** [comment]
@verygreen see also #889 please.
